realm/realm-library/src/main/java/io/realm/RealmObjectSchema.java

```diff
@@ -0,0 +1,511 @@
+     */
+    public RealmObjectSchema setNullable(String fieldName, boolean nullable) {
+        setRequired(fieldName, !nullable);
+        return this;
+    }
+
+    /**
+     * Checks if a given field is required, i.e. is not allowed to contain {@code null} values.
+     *
+     * @param fieldName field to check.
+     * @return {@code true} if it is requied, {@code false} otherwise.
+     * @throws IllegalArgumentException if field name doesn't exists.
+     * @see #setRequired(String, boolean)
+     */
+    public boolean isRequired(String fieldName) {
+        long columnIndex = table.getColumnIndex(fieldName);
+        return !table.isColumnNullable(columnIndex);
+    }
+
+    /**
+     * Checks if a given field is nullable, i.e. is allowed to contain {@code null} values.
+     *
+     * @param fieldName field to check.
+     * @return {@code true} if it is requied, {@code false} otherwise.
+     * @throws IllegalArgumentException if field name doesn't exists.
+     * @see #setNullable(String, boolean)
+     */
+    public boolean isNullable(String fieldName) {
+        long columnIndex = table.getColumnIndex(fieldName);
+        return table.isColumnNullable(columnIndex);
+    }
+
+    /**
+     * Checks if the class has a primary key defined.
+     *
+     * @return {@code true} if a primary key is defined, {@code false} otherwise.
+     * @see io.realm.annotations.PrimaryKey
+     */
+    public boolean hasPrimaryKey() {
+        return table.hasPrimaryKey();
+    }
+
+    /**
+     * Return all fields in this class.
+     *
+     * @return A list of all the fields in this class.
+     */
+    public Set<String> getFieldNames() {
+        int columnCount = (int) table.getColumnCount();
+        Set<String> columnNames = new TreeSet<String>();
+        for (int i = 0; i < columnCount; i++) {
+            columnNames.add(table.getColumnName(i));
+        }
+        return columnNames;
+    }
+
+    /**
+     * Runs a transformation on each RealmObject instance of the current class. The object will be represented as a
+     * {@link DynamicRealmObject}.
+     *
+     * @return This schema.
+     */
+    public RealmObjectSchema forEach(Transformer transformer) {
+        if (transformer != null) {
+            long size = table.size();
+            for (long i = 0; i < size; i++) {
+                transformer.apply(new DynamicRealmObject(realm, table.getCheckedRow(i)));
+            }
+        }
+
+        return this;
+    }
+
+    // Invariant: Field was just added
+    private void addModifiers(long columnIndex, FieldAttribute[] attributes) {
+        if (attributes != null && attributes.length > 0) {
+            if (containsAttribute(attributes, FieldAttribute.INDEXED)) {
+                table.addSearchIndex(columnIndex);
+            }
+
+            if (containsAttribute(attributes, FieldAttribute.PRIMARY_KEY)) {
+                table.setPrimaryKey(columnIndex);
+                table.addSearchIndex(columnIndex);
+            }
```

Comment:
well, `FieldAttribute.REQUIRED` is allowed to be passed in this function. but it is not handled.